2_process/src/calculate_toha_continuous.R

```diff
@@ -0,0 +1,232 @@
+  io <- mda.lakes::create_irr_day_cycle(lat,lon, dates=current_date, irr_mean = io, by='min')
+  
+  oha_df <- optical_habitat_area(io[[2]], kd, hypsos, irr_thresh[1], irr_thresh[2])
+  tha_df <- thermal_habitat_area(wtr_rmNA, hypsos, wtr_thresh[1], wtr_thresh[2])
+  
+  # Upsample tha (repeat values) to match the number of oha values (subdaily)
+  # Will even work if `nrow(wtr_rmNA) > 1`
+  tha_upsampled <- tha_df %>% slice(rep(1:n(), each = nrow(oha_df)))
+  toha <- thermal_optical_habitat_area(tha_upsampled, oha_df, hypsos)
+  
+  # Divide by number of timesteps to get average area per day
+  oha_daily <- sum(oha_df$habitat)/nrow(oha_df) 
+  tha_daily <- tha_df$habitat/nrow(tha_df) # should be the same before/after division since tha_df is already daily
+  toha_daily <- sum(toha)/length(toha)
+  
+  return(data.frame(opti_hab = oha_daily, 
+                    therm_hab = tha_daily, 
+                    opti_therm_hab = toha_daily))
+  
+}
+
+# Vectorized function
+optical_habitat_area <- function(I_0, Kd, hypso, I_lower, I_upper) {
+  
+  # Expecting only one Kd per day but if not, needs to match length of I_0
+  stopifnot(length(Kd) == 1 || length(Kd) == length(I_0))
+  
+  z_max <- tail(hypso$depths, 1)
+  z_surface <- head(hypso$depths, 1)
+  
+  ##### Find exact depths of irr thresholds 
+  
+  Z1 <- log( I_upper / I_0) / -Kd
+  Z2 <- log( I_lower / I_0) / -Kd
+  
+  ##### Checks before calculating areas
+  completely_below_lake <- which(Z1 > z_max) # if OHA is completely below lake, benthic area is 0 (too bright)
+  completely_above_lake <- which(I_0 < I_lower) # if OHA is completely above lake, benthic area is 0 (too dark)
+  benth_0 <- c(completely_below_lake, completely_above_lake)
+  
+  # Set Zs to NA so area calc is quicker (?)
+  Z1[benth_0] <- NA 
+  Z2[benth_0] <- NA
+  
+  # if OHA extends above lake, use the surface as Z1 to calc OHA
+  extends_above_lake <- which(Z1 < z_surface)
+  Z1[extends_above_lake] <- z_surface
+  
+  # if OHA extends below lake, use the bottom as Z2 to calc OHA
+  extends_below_lake <- which(Z2 >= z_max) # also include those with Z2 == z_max
+  Z2[extends_below_lake] <- z_max
+  
+  ##### Create hypso at these different depths
+  A1 <- resample_hypso(hypso, Z1)[["areas"]]
+  A2 <- resample_hypso(hypso, Z2)[["areas"]]
+  
+  ##### Use areas and depths at thresholds to calculate benthic area
+  
+  benthic_area <- calc_benthic_area(Z1, Z2, A1, A2)
+  
+  ##### Final benthic area checks
+  
+  # Set benthic areas to zero based on above criteria
+  benthic_area[completely_below_lake] <- 0 
+  benthic_area[completely_above_lake] <- 0
+  
+  # Add lake bottom if bottom of optical habitat 
+  benthic_area[extends_below_lake] <- benthic_area[extends_below_lake] + tail(hypso$areas, 1)
+  
+  return(data.frame(Z1 = Z1, Z2 = Z2, habitat = benthic_area))
+}
+
+# Now vectorized (except for one apply fxn)
+thermal_habitat_area <- function(wtr_df, hypso, wtr_lower, wtr_upper) {
+  
+  z_max <- tail(hypso$depths, 1)
```

Comment:
Before this line, do you want to `stopifnot(all(hypso$depths == cummax(hypso$depths)))`? This tests that depths are ordered correctly.

Reply:
Hmmm I did have code before that ordered them correctly before proceeding. Instead of stopping if they are out of order, it might make sense to order them in the `calculate_toha_per_lake` function?

Reply:
I think since this function _could_ be called alone, a quick check would make sense here.

Reply:
Realized that it does already get reordered in `calculate_toha_per_lake` so will add this stop to just `thermal_habitat` and `optical_habitat_area` to make sure `z_max` and `z_surface` are pulled correctly.